cmd/squid-per-site-exporter/main.go

```diff
@@ -0,0 +1,298 @@
+	mutex     sync.RWMutex
+	siteStats map[string]*SiteStats
+}
+
+func NewExporter() *Exporter {
+	return &Exporter{
+		siteStats: make(map[string]*SiteStats),
+	}
+}
+
+func (e *Exporter) parseLogLine(line string) {
+	// Squid log format: timestamp elapsedtime remotehost code/status bytes method URL rfc931 peerstatus/peerhost type
+	fields := strings.Fields(line)
+	if len(fields) < 7 {
+		return
+	}
+
+	// Extract relevant fields
+	elapsedTimeStr := fields[1]
+	codeStatus := fields[3]
+	bytesStr := fields[4]
+	method := fields[5]
+	urlStr := fields[6]
+
+	// Skip non-HTTP methods
+	if !strings.HasPrefix(method, "GET") && !strings.HasPrefix(method, "POST") &&
+		!strings.HasPrefix(method, "HEAD") && !strings.HasPrefix(method, "PUT") {
+		log.Printf("Unsupported method %q", method)
+		return
+	}
+
+	// Parse URL to extract hostname
+	parsedURL, err := url.Parse(urlStr)
+	if err != nil {
+		log.Printf("Invalid request URL %q: %v", urlStr, err)
+		return
+	}
+
+	hostname := parsedURL.Hostname()
+	if hostname == "" {
+		log.Printf("Missing hostname in URL %q", urlStr)
+		return
+	}
+
+	// Parse bytes
+	bytes, err := strconv.ParseInt(bytesStr, 10, 64)
+	if err != nil {
+		bytes = 0
+	}
+
+	// Parse elapsed time
+	elapsedTime, err := strconv.ParseFloat(elapsedTimeStr, 64)
+	if err != nil {
+		elapsedTime = 0
+	}
+
+	// Determine hit/miss from result code (token before '/')
+	// Consider only codes ending in "_HIT" as cache hits (e.g., TCP_HIT, MEM_HIT).
+	statusToken := codeStatus
+	if idx := strings.Index(codeStatus, "/"); idx >= 0 {
+		statusToken = codeStatus[:idx]
+	}
+	isHit := strings.HasSuffix(statusToken, "_HIT")
+
+	// Update stats
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+
+	if _, exists := e.siteStats[hostname]; !exists {
+		e.siteStats[hostname] = &SiteStats{}
+	}
+
+	stats := e.siteStats[hostname]
+	stats.Requests++
+	stats.Bytes += bytes
+
+	if isHit {
+		stats.Hits++
+	} else {
+		stats.Misses++
+	}
+
+	// Update Prometheus metrics
+	squidRequestsTotal.WithLabelValues(hostname).Inc()
+	squidBytesTotal.WithLabelValues(hostname).Add(float64(bytes))
+	squidResponseTime.WithLabelValues(hostname).Observe(elapsedTime / 1000.0) // Convert ms to seconds
+
+	if isHit {
+		squidHitTotal.WithLabelValues(hostname).Inc()
+	} else {
+		squidMissTotal.WithLabelValues(hostname).Inc()
+	}
+
+	// Ensure both hit and miss counters are initialized (even if 0) for this hostname
+	// This ensures squid_site_hits_total appears in metrics output even with 0 value
+	squidHitTotal.WithLabelValues(hostname).Add(0)
+	squidMissTotal.WithLabelValues(hostname).Add(0)
+
+	// Update hit ratio
+	if stats.Requests > 0 {
+		ratio := float64(stats.Hits) / float64(stats.Requests)
+		squidHitRatio.WithLabelValues(hostname).Set(ratio)
+	}
+}
+
+func (e *Exporter) readFromStdin() {
+	log.Printf("Reading squid logs from stdin")
+	scanner := bufio.NewScanner(os.Stdin)
+
+	for scanner.Scan() {
+		line := scanner.Text()
+		if line != "" {
+			e.parseLogLine(line)
```

Comment:
Maybe we can have something like this:

```suggestion
			e.parseFunc(line)
```

And then we can set `parseFunc = e.parseLogLine` when we initialize the exporter, while in tests we can set it to something that will allow us to test that parseFunc was called correctly.

This way, we can add tests for this function.

Reply:
updated